### ISSUE TITLE
Reproducible Builds

### ISSUE THREAD

**Issue by IzzySoft** [sweakpl/qralarm-android#46]
At IzzyOnDroid we support [Reproducible Builds](https://reproducible-builds.org/) (see: [Reproducible Builds, special client support and more in our repo](https://android.izzysoft.de/articles/named/iod-rbs-mirrors-clients)). Trying for yours, I was able to successfully generate the APK using `./gradlew assembleRelease`, but the resulting APKs were not identical. Was that APK really built from a clean tree at the commit the tag points to? If so, did I miss some build options? And if not, which commit was it? The APK diff is rather huge, here's the "head" of it:

```diff
--- /dev/fd/63  2024-07-11 16:17:45.952082322 +0200
+++ /dev/fd/62  2024-07-11 16:17:45.952082322 +0200
@@ -3,19 +3,27 @@
   META-INF/version-control-info.textproto
   32-bit CRC value (hex):                         54032ee8
   assets/dexopt/baseline.prof
-  32-bit CRC value (hex):                         5717ec58
+  32-bit CRC value (hex):                         d9484328
   assets/dexopt/baseline.profm
-  32-bit CRC value (hex):                         6e6f7f40
+  32-bit CRC value (hex):                         494da893
   classes.dex
-  32-bit CRC value (hex):                         ffd0e0df
+  32-bit CRC value (hex):                         16fa3f3f
   classes2.dex
   32-bit CRC value (hex):                         93669005
+  lib/arm64-v8a/libdatastore_shared_counter.so
+  32-bit CRC value (hex):                         4f56ac38
+  lib/armeabi-v7a/libdatastore_shared_counter.so
+  32-bit CRC value (hex):                         ffa6438b
+  lib/x86/libdatastore_shared_counter.so
+  32-bit CRC value (hex):                         db8d550b
+  lib/x86_64/libdatastore_shared_counter.so
+  32-bit CRC value (hex):                         54a56b50
   DebugProbesKt.bin
   32-bit CRC value (hex):                         d5ac4dc2
-  META-INF/services/N3.w
-  32-bit CRC value (hex):                         f7888b5c
-  META-INF/services/O3.a
-  32-bit CRC value (hex):                         f5ce3505
+  META-INF/services/Q3.w
+  32-bit CRC value (hex):                         e3a2b728
+  META-INF/services/R3.a
+  32-bit CRC value (hex):                         c88fe4eb
   kotlin-tooling-metadata.json
   32-bit CRC value (hex):                         89b4e17d
   kotlin/annotation/annotation.kotlin_builtins
@@ -65,7 +73,7 @@
...
```

Also strange that the APK I've built from the tagged commit has several `*.so` files missing in the APK attached to the release.

We'd appreciate if you could help making your build reproducible. We've prepared some [hints on reproducible builds](https://gitlab.com/IzzyOnDroid/repo/-/wikis/Reproducible-Builds) for that.

Looking forward to your reply!

cc @obfusk

**Comment by obfusk** [comment]
> Also strange that the APK I've built from the tagged commit has several `*.so` files missing in the APK attached to the release.

Looks like the APK was built before the datastore-preferences downgrade in 25992318c4e0107f97f7e894c89f9de5524b33c9?

**Comment by sweakpl** [comment]
Yes, that's probably right. First I've built and released the APK and later built the AAB and encountered this problem: https://issuetracker.google.com/issues/342671895

**Comment by IzzySoft** [comment]
Could you provide an APK built from a clean tree at the commit the latest tag points to? You can attach it to a comment here (simply *rename* it to `*.zip` – please do not put the APK into another zip). Then we could check if we can reproduce that, or if there might be other issues. Thanks in advance!

**Comment by sweakpl** [comment]
@IzzySoft here is the build from the latest commit:

[qralarm-android-release.zip](https://github.com/user-attachments/files/16677418/qralarm-android-release.zip)

**Comment by IzzySoft** [comment]
I see there's a new release out, so I've tried it. But oh, what did you build **that** from? According to `META-INF/version-control-info.textproto` from a commit with the hash 2ea3e91d2e3ed2f017b965401ba7659757d6afb2 – that doesn't even exist here. First basic rule from our above linked [hints on reproducible builds](https://gitlab.com/IzzyOnDroid/repo/-/wikis/Reproducible-Builds): always build the APK from a clean tree at the commit the tag points to. So: no chance to make 1.7.1 RB unfortunately :cry: Can we hope for a follow-up release respecting that rule? :pray:

**Comment by IzzySoft** [comment]
Can you please tell me how you build the app? I've just tried the latest release, and it's pretty close – but the dex is slightly different. I'm no Android dev, so I have to guess, could that be something with annotations?

```yaml
 |: const v14, #float 1.9141e+38 // #7f100043
-|: const v13, #float 1.91414e+38 // #7f1000f6
+|: const v13, #float 1.91414e+38 // #7f1000f4
 |: if-eqz v2, 024b // +0099
 |: const v2, #float 1.9141e+38 // #7f100040
 |: invoke-static {v2, v15}, Lj1/g;.v:(ILu0/o;)Ljava/lang/String;
@@ -15175,10 +15175,10 @@
 |: invoke-virtual {v15, v2}, Lu0/o;.Q:(I)V
 |: iget-boolean v2, v0, LL3/n;.d:Z
 |: if-eqz v2, 02ee // +0096
-|: const v2, #float 1.91413e+38 // #7f1000da
+|: const v2, #float 1.91413e+38 // #7f1000d8
 |: invoke-static {v2, v15}, Lj1/g;.v:(ILu0/o;)Ljava/lang/String;
 |: move-result-object v2
-|: const v3, #float 1.91413e+38 // #7f1000db
+|: const v3, #float 1.91413e+38 // #7f1000d9
 |: invoke-static {v3, v15}, Lj1/g;.v:(ILu0/o;)Ljava/lang/String;
```

The difference is always just the "comment" at the end (this is just an excerpt, there are more lines like that). In case it matters: I've tried with JDK-17 on Debian bookworm as well as with JDK-21 on Ubuntu jammy, identical results. I was just running `./gradlew assembleRelease`.

**Comment by sweakpl** [comment]
I've used JetBrains Runtime 21.0.3 (fork of OpenJDK) - this might have caused the issue.

**Comment by IzzySoft** [comment]
Yes, that could be. And should be no issue for your test/development builds, where those additional annotations might be helpful – but migght be one for release builds. Can you provide me an APK from the latest release tag, but compiled with OpenJDK so I can verify – and if that was the culprit, could you switch to OpenJDK for future release builds?

**Comment by IzzySoft** [comment]
Funny. That's identical to the APK attached to the 2.0.5 tag (same checksum), so not RB here. How did you build? I've built using `./gradlew assembleRelease` in a clean container on Debian-bookworm with JDK-17 as well as Ubuntu-jammy now with JDK-21 and could not reproduce. Both of my builds produced an identical APK as well, but not identical to yours. So something must be different in our build processes then. Could you please outline your build steps? Thanks in advance!

**Comment by sweakpl** [comment]
I am always building using the Android Studio IDE (Build -> Generate Signed App Bundle / APK...).

**Comment by IzzySoft** [comment]
Have you been running "Clean" first? I've never used Studio, but that's what I've been told is needed for a "clean build" there.

**Comment by sweakpl** [comment]
Ok, I might have missed the cleaning part - here is the "Clean" build:

[qralarm-android-signed-jdk-21-clean.zip](https://github.com/user-attachments/files/17923330/qralarm-android-signed-jdk-21-clean.zip)

If this one doesn't work, then there's something I have in mind that might be at fault here...

**Comment by IzzySoft** [comment]
Guess we have to go for that "something" then: the clean APK has the same checksum again…

**Comment by sweakpl** [comment]
Ok, my GitHub release workflow is like this:
* I push everything to GitHub,
* I build the release APK,
* I prepare a release on GitHub during which **I also create a tag** that is on the latest commit,
* I publish the release.

Now I suspect that the fact that the repository on GitHub has tags while my local repository does not can be causing those issues.

**Comment by IzzySoft** [comment]
> Now I suspect that the fact that the repository on GitHub has tags while my local repository does not can be causing those issues

No, not related (the tag is just a pointer, we both build from the commit). But can you please outline the second step: how do you **build** the APK? Well, basically you have: Android Studio, run "clean" and then "build". And Android Studio says it uses OpenJDK, right? Not Jetbrains or something else? Then I'm out of ideas :cry:

Some articles on checking/setting the JDK in Studio:

* [StackOverflow](https://stackoverflow.com/q/19608742/2533433)
* [official documentation](https://developer.android.com/build/jdks)

**Comment by sweakpl** [comment]
Yes, I'm pretty sure I am using OpenJDK as it is set as explained in the [documentation](https://developer.android.com/build/jdks#jdk-config-in-studio) you've linked:

![JDK location](https://github.com/user-attachments/assets/59bf3696-39ad-4760-b060-fee01755bb68)

With that setup, the docs are now assuring me that:

> If you run Gradle using the buttons in Android Studio, the JDK set in the Android Studio settings is used to run Gradle.

**Comment by IzzySoft** [comment]
Then I'm out of ideas where it might come from, and thus how to fix it. Your project directory has no local changes (`git status`, but also consider any `.gitignore` you might have)? All your builds from the same commit always have the same hash. So have mine. But your hash does not match my hash. Somewhere something must be different, but what?

Have you ever tried building at the command line? It's a straw, but if you'd go into your project directory and run `./gradlew assembleRelease`, and then `sha256sum *.apk` on the resulting APK, is that hash still identical to the one you've built before? Just trying to rule out some Android Studio settings I don't even know if they exist…

**Comment by IzzySoft** [comment]
> I think we've missed quite an important thing - my apk's are always signed with my .jks key. This is almost definitely causing this issue.

It would cause an issue if that was not the case – as the verifier needs to copy that signature on the APK it builds. You could have hidden something in the signing blocks we'd not cover otherwise.

But interesting already: so my "wild guess" it might have to do with Studio might not have been as far off as I thought. Otherwise, you should have the same hash again (and yes, I can confirm the first one from your previous builds).

I didn't keep the signed variant of my build, nor the logs, so I don't remember the hash of the APK with the copied key. Can you provide me that APK with the "new hash"? Then I can compare it directly here, as I've kept the unsigned APK of my build.

**Comment by IzzySoft** [comment]
> Can you explain what do you exactly mean by

You've mentioned:

> Hash for the apk built using `./gradlew assembleRelease` command: `D4F39CE51A04D71DC101B4315E431CDDEBC9432DE415C526D4E6D9FE7C85D3B9`

If I'd have the APK you refer to there, I could check if it's RB (i.e. if it matches the one I've built, after copying the signing block over).

**Comment by sweakpl** [comment]
Sure, here it is: [qralarm-android-unsigned.zip](https://github.com/user-attachments/files/17948717/qralarm-android-unsigned.zip)

**Comment by IzzySoft** [comment]
Ah, **unsigned** – OK, then I don't need to test: my **unsigned** APK has the hash `b823196e22ee6ae972eb228ef59945345e28fc1df03da09ca14db9f168ec7a0d`, which is different. But then, you've also built from a different commit: `0b2b7a6a0428b775ca7d6aeb1f0d606e3f659db7` according to the embedded commit hash. Let me try that one… Nope, also different hash (`9dac5afca7fc23d37b6400e75a4c3423618e5b7ebe64d4a92c57f9d26bdcfc44`). And the diff looks similar to the one before, so it's not that either.

So I'm out of ideas at this point. If you really have no uncommitted files in your tree, we're both using the same JDK (OpenJDK-21) on the same OS (my last runs have been on Ubuntu 22) with the same command (`./gradlew assembleRelease`) to create each the same APKs (hashes) each time from the same commit – but still have a different hash compared to each other, it's beyond me what might cause that. You've also used the gradle version defined here in your `gradle/wrapper/gradle-wrapper.properties`, right?

**Comment by sweakpl** [comment]
You've said:

> If you really have no uncommitted files in your tree

but what if I have a file that is .gitignored? `/src/main/res/values/secrets.xml` is ignored, but on my local machine it is always present when building. It is there because I have a different branch from which I am building the Google Play distribution and the file just stays there across checkouts.

Here is the apk built off of the commit `c27992ff92c43bfc1a7b4420ec8d108ab7bd8e44` without the `secrets.xml` file: [qralarm-android-unsigned.zip](https://github.com/user-attachments/files/17950395/qralarm-android-unsigned.zip)

I've checked its hash and it is: `E986C30B58C97F743982F7D9823417651071A745636205F0A3863788DDBD3F2B`.

**Comment by IzzySoft** [comment]
> but what if I have a file that is .gitignored?

You mean, what I asked you [2 days ago](https://github.com/sweakpl/qralarm-android/issues/46#issuecomment-2501899213)? Yes, that could be a cause. Can you e.g. clone the repo into a different directory and build from there?

> Here is the apk built off of the commit c27992ff92c43bfc1a7b4420ec8d108ab7bd8e44 without the secrets.xml file

Confirm it was built from that commit – but unfortunately, the diff is the same (except for the missing signature), so that was not it. But let's walk the differences:

* `res/rF.xml`: Linebreaks (you obviously build on Windows), can be fixed on our end.
* `META-INF/services/*`: same
* `assets/dexopt/baseline.prof`: differs as it contains the has from…
* `classes.dex`, we've looked at above. This is the only one we need to figure.

The dex diff is around 50 kB, so not exactly small. As I can only check the smali code here, it's not always clear what classes we talk about. Let me check if diffoscope gives something recognizable… Ah, I should have done that before. There are differences in `res/raw/keep.xml`:

![image](https://github.com/user-attachments/assets/b665e651-4491-4d79-908b-094079b5e80e)

Yuck. Looking at that file directly in your repo here, that seem to be line breaks again – so most likely a thing of "converting line breaks when checking in/out with git". All else are differences in smali code I can make nothing out of (that's beyond me). I can send you the files if you want to take a look at them: [diffs.zip](https://github.com/user-attachments/files/17951256/diffs.zip)

My guess is if you'd build on Linux (e.g. in a Podman or Docker container) the results would match. But apart from that, I don't know any solution. Hm, but maybe I could try something "dirty", like `unix2dos res/raw/keep.xml` before the build starts. I somehow doubt that would solve the other differences, though…

And indeed: diff got smaller, that `res/rF.xml` now matches. I suspect some other of your source files have a similar issue – but I cannot tell which ones, as in the Github repo here everything seems to have Unix line endings (I've checked out manually disabling autocrlf, and even the one file identified has them on Unix).

**Comment by IzzySoft** [comment]
I've just tried with the APK at v2.3.1, looks like the issue still remains. What are your plans here? Just asking for orientation, no pressure.

**Comment by sweakpl** [comment]
I think we've kind of hit the wall of the OS differences:

> My guess is if you'd build on Linux (e.g. in a Podman or Docker container) the results would match.

As for my plans on this issue, to put it lightly - it is not really on the top of my list of priorities, sorry.
Lately I'm iterating with new releases quite fast and I'm not really thinking about offloading the build process to some UNIX environment : (

**Comment by IzzySoft** [comment]
> I'm not really thinking about offloading the build process to some UNIX environment : (

Well, Github actions/workflows come to mind there, or using our [rbtlog](/obfusk/rbtlog) for the case (I could provide you with our recipe then).

But yeah, I get it if you're too busy on other tasks. Shall I put QRAlarm off our waiting list then, marking it as "not RB" – while you maybe keep this issue open as reminder and approach us then (e.g. with a ping here) when it "climbed up" your list of priorities?

**Comment by sweakpl** [comment]
> Shall I put QRAlarm off our waiting list then, marking it as "not RB"

Yeah, I think this would be a better idea.

**Comment by IzzySoft** [comment]
OK, done. Be welcome to ring my bell if you want to give it another try. Oh, and I leave it to you if you want to keep this issue open as reminder :wink:

Thanks for all your efforts and patience so far. I hope it brought some insights, and maybe bears some fruits later then. All the best!
